### PR TITLE
[libc++] Use correct size for deallocation of arrays in shared_ptr

### LINKED ISSUE #68051: std::allocate_shared & std::make_shared for arrays causes new-delete-size-mismatch under asan
The following snippet yields a new/delete type mismatch when building with sized deallocation:

```c++
int main(int argc, char *argv[]) {
    std::allocate_shared<int64_t[]>(std::allocator<int64_t>{}, 1);
}
```

It runs fine with `-stdlib=libc++ -fsanitize=address -std=c++20`, but fails when adding `-fsized-deallocation`. Asan stack trace:
```
=================================================================
==1==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x504000000050 in thread T0:
  object passed to delete has wrong type:
  size of the allocated type:   40 bytes;
  size of the deallocated type: 320 bytes.
    #0 0x561aa17a18d2 in operator delete(void*, unsigned long) /root/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:164:3
    #1 0x561aa17a4f6c in void std::__1::__libcpp_operator_delete[abi:v180000]<void*, unsigned long>(void*, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/new:282:3
    #2 0x561aa17a4f0c in void std::__1::__do_deallocate_handle_size[abi:v180000]<>(void*, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/new:308:10
    #3 0x561aa17a4eb4 in std::__1::__libcpp_deallocate[abi:v180000](void*, unsigned long, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/new:322:14
    #4 0x561aa17a4e39 in std::__1::allocator<std::__1::__sp_aligned_storage<8ul>>::deallocate[abi:v180000](std::__1::__sp_aligned_storage<8ul>*, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/allocator.h:130:13
    #5 0x561aa17a4df4 in std::__1::allocator_traits<std::__1::allocator<std::__1::__sp_aligned_storage<8ul>>>::deallocate[abi:v180000](std::__1::allocator<std::__1::__sp_aligned_storage<8ul>>&, std::__1::__sp_aligned_storage<8ul>*, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/allocator_traits.h:288:13
    #6 0x561aa17a4586 in std::__1::__unbounded_array_control_block<long [], std::__1::allocator<long>>::__on_zero_shared_weak() /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/shared_ptr.h:1140:9
    #7 0x561aa17a50ee in std::__1::__shared_weak_count::__release_shared[abi:v180000]() /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/shared_ptr.h:215:9
    #8 0x561aa17a31a5 in std::__1::shared_ptr<long []>::~shared_ptr[abi:v180000]() /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/shared_ptr.h:775:23
    #9 0x561aa17a305b in main /app/example.cpp:5:5
    #10 0x7f028f312082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #11 0x561aa16c536d in _start (/app/output.s+0x2c36d)

0x504000000050 is located 0 bytes inside of 40-byte region [0x504000000050,0x504000000078)
allocated by thread T0 here:
    #0 0x561aa17a0c6d in operator new(unsigned long) /root/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:95:3
    #1 0x561aa17a3af4 in void* std::__1::__libcpp_operator_new[abi:v180000]<unsigned long>(unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/new:272:10
    #2 0x561aa17a3a5c in std::__1::__libcpp_allocate[abi:v180000](unsigned long, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/new:298:10
    #3 0x561aa17a399c in std::__1::allocator<std::__1::__sp_aligned_storage<8ul>>::allocate[abi:v180000](unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/allocator.h:114:38
    #4 0x561aa17a392c in std::__1::allocator_traits<std::__1::allocator<std::__1::__sp_aligned_storage<8ul>>>::allocate[abi:v180000](std::__1::allocator<std::__1::__sp_aligned_storage<8ul>>&, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/allocator_traits.h:268:20
    #5 0x561aa17a3626 in std::__1::__allocation_guard<std::__1::allocator<std::__1::__sp_aligned_storage<8ul>>>::__allocation_guard[abi:v180000]<std::__1::allocator<long>>(std::__1::allocator<long>, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/allocation_guard.h:57:18
    #6 0x561aa17a3317 in std::__1::shared_ptr<long []> std::__1::__allocate_shared_unbounded_array[abi:v180000]<long [], std::__1::allocator<long>>(std::__1::allocator<long> const&, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/shared_ptr.h:1162:39
    #7 0x561aa17a30f7 in std::__1::shared_ptr<long []> std::__1::allocate_shared[abi:v180000]<long [], std::__1::allocator<long>, void>(std::__1::allocator<long> const&, unsigned long) /opt/compiler-explorer/clang-trunk-20231002/bin/../include/c++/v1/__memory/shared_ptr.h:1301:12
    #8 0x561aa17a3052 in main /app/example.cpp:5:5
    #9 0x7f028f312082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

SUMMARY: AddressSanitizer: new-delete-type-mismatch /root/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:164:3 in operator delete(void*, unsigned long)
```

Live snippet: https://godbolt.org/z/ddYr6TYb1

Note: the mismatch was found by running the [`std::boyer_moore_searcher` example snippet on cppreference](https://en.cppreference.com/w/cpp/utility/functional/boyer_moore_searcher): https://godbolt.org/z/Gsx6dWPYn. Likely it has the same underlying issue, although this one does not require c++20 to reproduce.

The [allocation](https://github.com/llvm/llvm-project/blob/824b1677a44e25b7c9808c774ba2d894ff14df2b/libcxx/include/__memory/shared_ptr.h#L1162) divides by `sizeof(_AlignedStorage)`, but the [deallocation](https://github.com/llvm/llvm-project/blob/824b1677a44e25b7c9808c774ba2d894ff14df2b/libcxx/include/__memory/shared_ptr.h#L1140) does not appear to make the same adjustment. That may be the source of the issue, which could be fixed w/ something like:

```diff
--- a/libcxx/include/__memory/shared_ptr.h
+++ b/libcxx/include/__memory/shared_ptr.h
@@ -1137,7 +1137,9 @@
         __alloc_.~_Alloc();
         size_t __size = __unbounded_array_control_block::__bytes_for(__count_);
         _AlignedStorage* __storage = reinterpret_cast<_AlignedStorage*>(this);
-        allocator_traits<_StorageAlloc>::deallocate(__tmp, _PointerTraits::pointer_to(*__storage), __size);
+        allocator_traits<_StorageAlloc>::deallocate(
+            __tmp, _PointerTraits::pointer_to(*__storage),
+            __size / sizeof(_AlignedStorage));
     }
```

### PR DESCRIPTION
**Pull request by ilya-biryukov**
Fixes #68051.

Current implementation passes the number of `_AlignedStorage` objects when it calls to `allocate` and the number of **bytes** on deallocate. This only applies to allocations that allocate control block and the storage together, i.e. `make_shared` and `allocate_shared`.

Found by ASan.

### CODE REVIEW

**Comment by EricWF** [comment]
I wonder why ASAN wasn't flagging this externally. Any clues?

**Comment by EricWF** [comment]
Could you fix the missing include in the failing test, then I think this is good to go!

Though it would be nice if we could figure out why ASAN wasn't firing here.

**Review by EricWF — APPROVED** [review]
LGTM after getting the test passing (not your failure, but lets get the CI green).

File: libcxx/test/libcxx/memory/shared_ptr_array.pass.cpp
Comment (ldionne) [inline libcxx/test/libcxx/memory/shared_ptr_array.pass.cpp:19]: This needs to be `int main(int, char**)`.

File: libcxx/test/libcxx/memory/shared_ptr_array.pass.cpp
Comment (ldionne) [inline libcxx/test/libcxx/memory/shared_ptr_array.pass.cpp:25]: And we need `return 0;` here.

Otherwise this will fail if we compile with `-ffreestanding`.

File: libcxx/include/__memory/shared_ptr.h
Comment (ldionne) [inline libcxx/include/__memory/shared_ptr.h:1141]: Is this not always `__count_ + 1`? i.e. the number of elements in the array + the size of the control block at the beginning? I guess I'm kinda confused with my own code now.

**Review by ldionne — COMMENTED** [review]
Thanks for the fix! This is indeed a nasty bug, a bit surprising that it went unnoticed for so long.

File: libcxx/test/libcxx/memory/shared_ptr_array.pass.cpp
Comment (ilya-biryukov) [reply libcxx/test/libcxx/memory/shared_ptr_array.pass.cpp]: Done.

File: libcxx/test/libcxx/memory/shared_ptr_array.pass.cpp
Comment (ilya-biryukov) [reply libcxx/test/libcxx/memory/shared_ptr_array.pass.cpp:27]: Done.

**Comment by ilya-biryukov** [comment]
> I wonder why ASAN wasn't flagging this externally. Any clues?


> Thanks for the fix! This is indeed a nasty bug, a bit surprising that it went unnoticed for so long.

I think it's because it requires a rather rare combination of flags. First, we need to enable `-fsized-deallocation`[1] and build with ASan. In addition, one either needs to use `std::boyer_moore_searcher` in C++17 (which is a likely a rarely used API) or  C++20's `std::make_shared` with array types (which should be more common, but C++20 is relatively rare itself).

[1]: it is surprising that `-fsized-deallocation` is not the default yet given that it was part of C++14. There was an [attempt](https://reviews.llvm.org/D112921#4624744) to make it the default in August 2023, but it got reverted.

**Comment by ilya-biryukov** [comment]
I'm going to merge the change in as all comments are addressed and this already has LGTM from @EricWF.
The only pending build is for FreeBSD and it came out green last time (there were no significant changes afterwards), the change itself seems relatively straightforward.

I hope I don't break any policies or expectations here, but please let me know if that's the case.

File: libcxx/include/__memory/shared_ptr.h
Comment (sxeraverx) [reply libcxx/include/__memory/shared_ptr.h:1141]: If `alignof(size_t) < sizeof(size_t)`, it will not be `__count_ + 1`.

**Comment by jyknight** [comment]
Any objection to backporting this to 17.x branch?

**Comment by ldionne** [comment]
No, we should totally do that. Do you want to create the cherry-pick issue?

**Pull request merged**
